akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -0,0 +1,379 @@
+
+  override def carbonCopy: Module = copy()
+
+  override def replaceShape(s: Shape): Module = ???
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] object ActorGraphInterpreter {
+  trait BoundaryEvent extends DeadLetterSuppression with NoSerializationVerificationNeeded
+
+  final case class OnError(id: Int, cause: Throwable) extends BoundaryEvent
+  final case class OnComplete(id: Int) extends BoundaryEvent
+  final case class OnNext(id: Int, e: Any) extends BoundaryEvent
+  final case class OnSubscribe(id: Int, subscription: Subscription) extends BoundaryEvent
+
+  final case class RequestMore(id: Int, demand: Long) extends BoundaryEvent
+  final case class Cancel(id: Int) extends BoundaryEvent
+  final case class SubscribePending(id: Int) extends BoundaryEvent
+  final case class ExposedPublisher(id: Int, publisher: ActorPublisher[Any]) extends BoundaryEvent
+
+  case object Resume extends BoundaryEvent
+
+  final class BoundarySubscription(val parent: ActorRef, val id: Int) extends Subscription {
+    override def request(elements: Long): Unit = parent ! RequestMore(id, elements)
+    override def cancel(): Unit = parent ! Cancel(id)
+    override def toString = "BoundarySubscription" + System.identityHashCode(this)
+  }
+
+  final class BoundarySubscriber(val parent: ActorRef, id: Int) extends Subscriber[Any] {
+    override def onError(cause: Throwable): Unit = {
+      ReactiveStreamsCompliance.requireNonNullException(cause)
+      parent ! OnError(id, cause)
+    }
+    override def onComplete(): Unit = parent ! OnComplete(id)
+    override def onNext(element: Any): Unit = {
+      ReactiveStreamsCompliance.requireNonNullElement(element)
+      parent ! OnNext(id, element)
+    }
+    override def onSubscribe(subscription: Subscription): Unit = {
+      ReactiveStreamsCompliance.requireNonNullSubscription(subscription)
+      parent ! OnSubscribe(id, subscription)
+    }
+  }
+
+  def props(assembly: GraphAssembly, shape: Shape, settings: ActorMaterializerSettings): Props =
+    Props(new ActorGraphInterpreter(assembly, shape, settings)).withDeploy(Deploy.local)
+
+  class BatchingActorInputBoundary(size: Int, id: Int) extends UpstreamBoundaryStageLogic[Any] {
+    require(size > 0, "buffer size cannot be zero")
+    require((size & (size - 1)) == 0, "buffer size must be a power of two")
+
+    private var upstream: Subscription = _
+    private val inputBuffer = Array.ofDim[AnyRef](size)
+    private var inputBufferElements = 0
+    private var nextInputElementCursor = 0
+    private var upstreamCompleted = false
+    private var downstreamCanceled = false
+    private val IndexMask = size - 1
+
+    private def requestBatchSize = math.max(1, inputBuffer.length / 2)
+    private var batchRemaining = requestBatchSize
+
+    val out: Outlet[Any] = Outlet[Any]("UpstreamBoundary" + id)
+
+    private def dequeue(): Any = {
+      val elem = inputBuffer(nextInputElementCursor)
+      assert(elem ne null)
```

Comment:
I’d even say that we should flat-out ban `assert` from the code base outside of tests.